polygeist/lib/polygeist/Passes/AffineReduction.cpp

```diff
@@ -1,264 +1,367 @@
-        newOperands.push_back(std::get<1>(pair)->getOperand(0));
-        // rewriter.eraseOp(std::get<1>(pair));
-      }
-      mergedTerminator.getOperandsMutable().assign(newOperands);
-    };
-
-    auto mergedYieldOp = cast<AffineYieldOp>(newBlock->getTerminator());
-    cloneFilteredTerminator(mergedYieldOp);
-
-    // prepare for new yielded value for 'replaceOp'.
-    SmallVector<Value, 4> newYieldedRes;
-    SmallVector<Value, 4> newRes(newForOp.getResults());
-    int additionalRes =
-        newForOp.getResults().size() - forOp.getResults().size();
-    assert(additionalRes >= 0 && "must be >= 0");
-    newRes.insert(newRes.end(), newRes.begin(), newRes.end() - additionalRes);
-
-    // propagate results new forOp to downstream loads if any,
-    // otherwise insert a store right after the for. The stored
-    // element is the result of the for.
-    assert(candidateOpsInFor.size() == loadsInFor.size());
-    i = 0;
-
-    DominanceInfo DT;
-    PostDominanceInfo PDT;
-    for (auto pair : candidateOpsInFor) {
-      auto store = cast<AffineStoreOp>(std::get<1>(pair));
-
-      auto loads = loadsInFor[i];
-      for (auto *load : loads) {
-        if (PDT.postDominates(store, load)) {
-          load->getResult(0).replaceAllUsesWith(
-              newForOp.getBody()->getArguments()[i + origNumRegionArgs + 1]);
-        } else if (DT.dominates(store, load)) {
-          load->getResult(0).replaceAllUsesWith(store.getOperand(0));
-        } else {
-
-          assert(0 && "illegal behavior");
+    Rewriter.mergeBlocks(OldBlock, NewBlock, NewBlockTransferArgs);
+
+    auto CloneFilteredTerminator =
+        [ReductionOps](AffineYieldOp MergedTerminator) {
+          SmallVector<Value, 4> NewOperands;
+          llvm::append_range(NewOperands, MergedTerminator.getOperands());
+          // store operands are now returned.
+          for (const ReductionOp &Op : ReductionOps)
+            NewOperands.push_back(Op.getStore()->getOperand(0));
+          MergedTerminator.getOperandsMutable().assign(NewOperands);
+        };
+
+    auto MergedYieldOp = cast<AffineYieldOp>(NewBlock->getTerminator());
+    CloneFilteredTerminator(MergedYieldOp);
+
+    // Prepare for new yielded value for 'replaceOp'.
+    SmallVector<Value, 4> NewYieldedRes, NewRes(NewForOp.getResults());
+    size_t AdditionalRes =
+        NewForOp.getResults().size() - ForOp.getResults().size();
+    assert(AdditionalRes >= 0 && "must be >= 0");
+    NewRes.insert(NewRes.end(), NewRes.begin(), NewRes.end() - AdditionalRes);
+
+    // Propagate results new forOp to downstream loads if any, otherwise insert
+    // a store right after the for. The stored element is the result of the for.
+    {
+      DominanceInfo DT;
+      PostDominanceInfo PDT;
+      size_t ArgNo = 0;
+      [[maybe_unused]] SmallVector<AffineStoreOp> NewStores;
+
+      for (ReductionOp &Op : ReductionOps) {
+        for (Operation *Load : Op.getOtherLoads()) {
```

Comment:
```suggestion
        for (AffineLoadOp Load : Op.getOtherLoads()) {
```
Use `AffineLoadOp` member functions below instead of generic ones.